### PR TITLE
Leiden C++ API

### PR DESCRIPTION
**Pull request by naimnv**
#2485 
API definition for Leiden C++ Implementation

### CODE REVIEW

File: cpp/src/community/leiden_sg.cu [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
```
Comment (seunghwak) [inline cpp/src/community/leiden_sg.cu:2]: 2021 => 2022

File: cpp/src/community/leiden_impl.cuh [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+ * limitations under the License.
+ */
+#pragma once
+
+#include <prims/edge_partition_src_dst_property.cuh>
+#include <prims/reduce_op.cuh>
+#include <prims/transform_reduce_v_frontier_outgoing_e_by_dst.cuh>
+#include <prims/update_edge_partition_src_dst_property.cuh>
+#include <prims/update_v_frontier.cuh>
+#include <prims/vertex_frontier.cuh>
```
Comment (seunghwak) [inline cpp/src/community/leiden_impl.cuh:23]: Don't forget to delete irrelevant headers.

File: cpp/src/community/leiden_impl.cuh [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+#include <limits>
+#include <type_traits>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename GraphViewType>
+void check_clustering(GraphViewType const& graph_view, typename GraphViewType::vertex_t* clustering)
```
Comment (seunghwak) [inline cpp/src/community/leiden_impl.cuh:54]: I assume `check_clustering` and `flatten_dendrogram` are copied from the louvain implementation. We'd better not replicate code. Need to think about proper refactoring. Maybe add a FIXME statement (`//FIXME: what should be fixed...`) if you plan to figure this out later.

File: cpp/src/community/leiden_impl.cuh [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                   vertex_ids_v.begin(),
+                   vertex_ids_v.end(),
+                   graph_view.local_vertex_partition_range_first());
+  partition_at_level<vertex_t, multi_gpu>(
+    handle, dendrogram, vertex_ids_v.data(), clustering, dendrogram.num_levels());
+}
+
+}  // namespace detail
+
+// Keep the API similar to Louvain
```
Comment (seunghwak) [inline cpp/src/community/leiden_impl.cuh:105]: I guess this comment is unnecessary.

File: cpp/src/community/leiden_impl.cuh (hunk 2 above)
Comment (naimnv) [reply cpp/src/community/leiden_impl.cuh]: Indeed.  Removed.

**Review by ChuckHastings — COMMENTED** [review]
If you put:
`closes #xxxx` in the description then when the PR is merged it will automatically close the issue.

File: cpp/src/community/leiden_impl.cuh [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+template <typename GraphViewType>
+void flatten_dendrogram(raft::handle_t const& handle,
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/src/community/leiden_impl.cuh:90]: Is it due to upstream compatibility (i.e. for C and then python api) ?

[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/src/community/leiden_impl.cuh:90]: Yes, but now we have pylibcugraph and the C API.  We can make a C API that is backwards compatible without needing to keep the C++ API backwards compatible.  The python code will need to be rewritten anyway, so we might not even worry about that.

File: cpp/src/community/leiden_impl.cuh [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+          typename GraphViewType::weight_type>
+leiden(raft::handle_t const& handle,
+       GraphViewType const& graph_view,
+       size_t max_level,
+       typename GraphViewType::weight_type resolution)
+{
+  return detail::leiden(handle, graph_view, max_level, resolution);
+}
+
+template <typename GraphViewType>
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/src/community/leiden_impl.cuh:103]: Resolved.

File: cpp/src/community/leiden_impl.cuh (hunk 5 above)
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/src/community/leiden_impl.cuh]: Resolved.

File: cpp/src/community/leiden_impl.cuh [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+#include <thrust/tuple.h>
+
+#include <limits>
+#include <type_traits>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename GraphViewType>
```
[the remark this answers is not in the document]
Comment (naimnv) [reply cpp/src/community/leiden_impl.cuh]: I have changed it to 
```
template <typename vertex_t,
          typename edge_t,
          typename weight_t,
          bool multi_gpu,
          bool store_transposed = false>
```

**Review by ChuckHastings — APPROVED** [review]

**Review by seunghwak — APPROVED** [review]

**Comment by naimnv** [comment]
@seunghwak  

`Merging is blocked
The base branch restricts merging to authorized users. [Learn more about protected branches.](https://docs.github.com/articles/about-protected-branches/)

You’re not [authorized](https://docs.github.com/articles/about-protected-branches/) to merge this pull request.`

As I'm not authorized to merge, could you merge it or give me access?

**Comment by seunghwak** [comment]
@naimnv Chuck or Brad will merge the approved PRs.

**Comment by ChuckHastings** [comment]
@gpucibot merge

**Pull request merged**